### PR TITLE
[WIP] Pod networking proxy

### PR DESCRIPTION
**Pull request by vladikr**
WIP

Re-work networking codebase to allow attaching various interface types
Introduce a proxy mode that will allow users to bind to a specific VM ports 

```yaml
kind: VM
spec:
  devices:
    interfaces:
      - name: nic-ip
        networkName: red
        proxy: ## EITHER THIS
          ports:
          - name: ssh
        bridge: ## OR THIS
          delegateIp: true # offers the ip in case that the source has an ip
        Macvtap: ## would error with pod
          mode: vepa
  networks:
  - name: red
    pod: {} # Stock pod network
```

### CODE REVIEW

**Comment by rmohr** [comment]
@vladikr could you explain what you want to achive? Which functionality is meant for the pod network, which functionality for completely separate nics?

File: pkg/virt-launcher/virtwrap/network/proxyinterface.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	preparePodNetworkInterfaces(nicLink netlink.Link) error
+	startDHCPServer()
+	decorateInterfaceConfig() *api.Interface
+}
+
+type PodProxyInterface struct{}
+
+func (l *PodProxyInterface) Unplug() {}
+
+// SetupDefaultPodNetwork will prepare the pod management network to be used by a virtual machine
```
Comment (phoracek) [inline pkg/virt-launcher/virtwrap/network/proxyinterface.go]: Outdated comment

File: pkg/virt-launcher/virtwrap/network/proxyinterface.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+// which will own the pod network IP and MAC. Pods MAC address will be changed to a
+// random address and IP will be deleted. This will also create a macvlan device with a fake IP.
+// DHCP server will be started and bounded to the macvlan interface to server the original pod ip
+// to the guest OS
+func (l *PodProxyInterface) Plug(iface *v1.Interface, domain *api.Domain) error {
+	precond.MustNotBeNil(domain)
```
Comment (phoracek) [inline pkg/virt-launcher/virtwrap/network/proxyinterface.go]: The same should apply for iface?

File: pkg/virt-launcher/virtwrap/network/proxyinterface.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+		// After the network is configured, cache the result
+		// in case this function is called again.
+		ifconf = driver.decorateInterfaceConfig()
+		err = setCachedInterface(ifconf)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// TODO:(vladikr) Currently we support only one interface per vm.
```
Comment (phoracek) [inline pkg/virt-launcher/virtwrap/network/proxyinterface.go]: Is still still true?

File: pkg/virt-launcher/virtwrap/network/proxyinterface.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+	}
+
+	if ifconf == nil {
+		podNicLink, err := driver.discoverPodNetworkInterface()
+		if err != nil {
+			return err
+		}
+
+		if err := driver.preparePodNetworkInterfaces(podNicLink); err != nil {
+			log.Log.Reason(err).Critical("failed to prepared pod networking")
```
Comment (phoracek) [inline pkg/virt-launcher/virtwrap/network/proxyinterface.go]: s/prepared/prepare

File: pkg/virt-launcher/virtwrap/network/proxyinterface.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+func (b *IPBinding) startDHCPServer() {
+	// Start DHCP Server
+	fakeServerAddr, _ := netlink.ParseAddr(bridgeFakeIP)
+	go Handler.StartDHCP(b.vif, fakeServerAddr)
+}
+
+func (b *IPBinding) decorateInterfaceConfig() *api.Interface {
+
```
Comment (phoracek) [inline pkg/virt-launcher/virtwrap/network/proxyinterface.go]: This empty line could be dropped

File: pkg/virt-launcher/virtwrap/network/proxyinterface.go [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+	inter := api.Interface{}
```
Comment (phoracek) [inline pkg/virt-launcher/virtwrap/network/proxyinterface.go]: Or
```go
inter := api.Interface{
    Type:   "bridge",
    Source: api.InterfaceSource{Bridge: bridgeName},
    MAC:    &api.MAC{MAC: b.vif.MAC.String()},
    Model:  &api.Model{Type: "virtio"},
}
```

File: pkg/virt-launcher/virtwrap/network/proxyinterface.go [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+		return nil, err
+	}
+	b.vif.MAC = mac
+	return nicLink, nil
+}
+
+func (b *IPBinding) preparePodNetworkInterfaces(nicLink netlink.Link) error {
+	// Remove IP from POD interface
+	err := Handler.AddrDel(nicLink, &b.vif.IP)
+
```
Comment (phoracek) [inline pkg/virt-launcher/virtwrap/network/proxyinterface.go]: This empty line could be dropped

File: api/openapi-spec/swagger.json [hunk 8]
```diff
@@ -3583,6 +3620,57 @@
...
+    "properties": {
+     "bridge": {
+      "$ref": "#/definitions/v1.InterfaceBridge"
+     },
+     "name": {
+      "description": "Name of the interface",
+      "type": "string"
+     },
+     "networkName": {
+      "description": "Name of the refernced networks\nMust match the Name of a Network",
```
Comment (phoracek) [inline api/openapi-spec/swagger.json]: referenced

File: api/openapi-spec/swagger.json [hunk 9]
```diff
@@ -3680,6 +3768,21 @@
      }
     }
    },
+   "v1.Network": {
+    "description": "Network represents a network type and an resource that should be connected to the vm.",
```
Comment (phoracek) [inline api/openapi-spec/swagger.json]: s/an/a

File: api/openapi-spec/swagger.json [hunk 10]
```diff
@@ -3239,6 +3239,36 @@
...
+      "description": "Number of port to expose on the pod's IP address. This must be a valid port number, 0 \u003c x \u003c 65536.",
+      "type": "integer",
+      "format": "int32"
+     },
+     "hostIP": {
+      "description": "What host IP to bind the external port to.",
+      "type": "string"
+     },
+     "hostPort": {
+      "description": "Number of port to expose on the host. If specified, this must be a valid port number, 0 \u003c x \u003c 65536. If HostNetwork is specified, this must match ContainerPort. Most containers do not need this.",
```
[the remark this answers is not in the document]
Comment (fabiand) [reply api/openapi-spec/swagger.json]: Looks like a copy nad paste? ;)

[the remark this answers is not in the document]
Comment (vladikr) [reply api/openapi-spec/swagger.json]: Yea, at the moment... it's wip, I'll adjust this for vms :)

**Comment by vladikr** [comment]
some progress. not there yet :)

**Comment by rmohr** [comment]
Dump question: As far as I understand it, we are introducing this to work with frameworks like istio? Are we sure that our current approach does not integrate into these frameworks?

**Pull request closed without merging**